### PR TITLE
Add type hinting and tactile detect prevention

### PR DESCRIPTION
**Pull request by georgiablanco**
## Proposed changes

Add type hinting and tactile detect prevention when the robot is running a demo

## Checklist

Before posting a PR ensure that from each of the below categories **AT LEAST ONE BOX HAS BEEN CHECKED**. If more than one category is applicable then more can be checked. Also ensure that the proposed changes have been filled out with relevant information for reviewers.

## Tests

- [ ] No tests required to be added. (For small changes that will be tested by CI/CD infrastructure).
- [ ] Added/Modified automated and PhantomHand CI tests (if a new class is added (Python or C++), the interface of that class must be unit tested).
- [x] Manually tested in simulation (if simulation specific or no hardware required to test the functionality). 
- [x] Manually tested on hardware (if hardware specific or related).

## Documentation

- [x] No documentation required to be added.
- [ ] Added documentation (For any new feature, explain what it does and how to use it. Write the documentation in a relevant space, e.g. Github, Confluence, etc).
- [ ] Updated documentation (For changes to pre-existing features mentioned in the documentation).

### CODE REVIEW

File: sr_demos/scripts/sr_demos/demo.py [hunk 1]
```diff
@@ -225,24 +230,27 @@ def _has_tactiles(self):
...
-    def _get_joint_states_for_robot(self, joint_states_config_yaml):
+    def _get_joint_states_for_robot(self,
+                                    joint_states_config_yaml: Dict[str, Dict[str, int]]) -> Dict[str, Dict[str, int]]:
         '''
             Returns a dictionary of joint states with correct prefix
             and only joints that are present in the hand type
-            Input: joint_states_config_yaml: dictionary of joint states
-            Output: joint_states: dictionary of joint states
+            @param  joint_states_config_yaml: dictionary of joint states
+            @return joint_states: dictionary of joint states
```
Comment (bdenoun) [inline sr_demos/scripts/sr_demos/demo.py]: I don't think the return should have a named parameter after it
```suggestion
            @return: dictionary of joint states
```

File: sr_demos/scripts/sr_demos/demo.py [hunk 2]
```diff
@@ -225,24 +230,27 @@ def _has_tactiles(self):
...
+    def _correct_joint_states_for_hand_type(self,
+                                            joint_states_config: Dict[str, Dict[str, int]]) -> Dict[str, Dict[str,
+                                                                                                              int]]:
         '''
             Returns a dictionary of joint states for the robot type
             where the joints that are not present in the hand type are removed
-            Input: joint_states_config: dictionary of joint states
-            Output: joint_states_config: dictionary of joint states
+            @param joint_states_config: dictionary of joint states
+            @return joint_states_config: dictionary of joint states
```
Comment (bdenoun) [inline sr_demos/scripts/sr_demos/demo.py]:
```suggestion
            @return: dictionary of joint states
```

File: sr_demos/scripts/sr_demos/demo.py [hunk 3]
```diff
@@ -255,15 +263,15 @@ def _correct_joint_states_for_hand_type(self, joint_states_config):
...
         return joint_states_config
 
-    def _add_prefix_to_joint_states(self, joint_states_config):
+    def _add_prefix_to_joint_states(self, joint_states_config: Dict[str, Dict[str, int]]) -> Dict[str, Dict[str, int]]:
         '''
             Adds the correct prefix to the joint states
-            Input: joint_states_config - dictionary of joint states
-            Output: demo_states - dictionary of joint states with correct prefix for the robot type
+            @param joint_states_config - dictionary of joint states
+            @return demo_states - dictionary of joint states with correct prefix for the robot type
```
Comment (bdenoun) [inline sr_demos/scripts/sr_demos/demo.py]:
```suggestion
            @return: dictionary of joint states with correct prefix for the robot type
```

File: sr_demos/scripts/sr_demos/demo.py [hunk 4]
```diff
@@ -316,10 +325,28 @@ def check_touched_finger(self):
             touched_finger = self.tactiles[0].confirm_touched()
         return touched_finger
 
+    def _enable_touch_detection(self, detect: bool):
```
Comment (bdenoun) [inline sr_demos/scripts/sr_demos/demo.py]: I think you are missing the output hint
```suggestion
    def _enable_touch_detection(self, detect: bool) -> None:
```

File: sr_demos/scripts/sr_demos/demo.py [hunk 5]
```diff
@@ -316,10 +325,28 @@ def check_touched_finger(self):
...
+            tactile.touch_detection = detect
+
+    def touch_detection_enabled(self) -> bool:
+        '''
+            Returns whether touch detection is enabled or not
+            @return True or False
+        '''
+        if not any(tactile.touch_detection for tactile in self.tactiles):
+            return False
+        return True
```
Comment (bdenoun) [inline sr_demos/scripts/sr_demos/demo.py]:
```suggestion
        return any(tactile.touch_detection for tactile in self.tactiles)
```

Comment (bdenoun) [reply sr_demos/scripts/sr_demos/demo.py]: Much simpler and more readable :)

File: sr_demos/scripts/sr_demos/demo.py [hunk 6]
```diff
@@ -316,10 +325,28 @@ def check_touched_finger(self):
...
+
     def stored_states_sequence(self):
```
Comment (bdenoun) [inline sr_demos/scripts/sr_demos/demo.py]: No type hinting?

File: sr_demos/scripts/sr_demos/demo.py [hunk 7]
```diff
@@ -352,8 +379,13 @@ def stored_states_sequence(self):
 
         self.commander.run_named_trajectory(trajectory)
         rospy.loginfo("Stored States demo completed")
+        self._enable_touch_detection(True)
 
     def standard_demo_sequence(self):
```
Comment (bdenoun) [inline sr_demos/scripts/sr_demos/demo.py]: No type hinting?

File: sr_demos/scripts/sr_demos/demo.py [hunk 8]
```diff
@@ -405,11 +437,13 @@ def standard_demo_sequence(self):
         self.execute_command_check('zero_wr', 0.4)
         self.commander.move_to_named_target("open")
         rospy.loginfo("Standard demo completed")
+        self._enable_touch_detection(True)
 
     def rock_paper_scissors(self):
```
Comment (bdenoun) [inline sr_demos/scripts/sr_demos/demo.py]: No type hinting?

File: sr_demos/scripts/sr_demos/demo.py [hunk 9]
```diff
@@ -441,11 +475,13 @@ def rock_paper_scissors(self):
             self.commander.move_to_named_target("fingers_pack_thumb_open")
         self.commander.move_to_named_target("open")
         rospy.loginfo("Rock, Paper, Scissors demo completed")
+        self._enable_touch_detection(True)
 
     def grasp_demo(self):
```
Comment (bdenoun) [inline sr_demos/scripts/sr_demos/demo.py]: No type hinting?

File: sr_demos/scripts/sr_demos/demo.py [hunk 10]
```diff
@@ -491,6 +527,7 @@ def grasp_demo(self):
         self.commander.move_to_named_target("open")
 
         rospy.loginfo("Grasp Demo completed")
+        self._enable_touch_detection(True)
 
     def complete_random_sequence(self):
```
Comment (bdenoun) [inline sr_demos/scripts/sr_demos/demo.py]: No type hinting?

**Review by bdenoun — CHANGES_REQUESTED** [review]

File: sr_demos/scripts/sr_demos/demo.py (hunk 6 above)
Comment (georgiablanco) [reply sr_demos/scripts/sr_demos/demo.py]: fixed in https://github.com/shadow-robot/sr_interface/pull/823/commits/4d899740e14c6a37b20146403ea06d3b82a837c8

File: sr_demos/scripts/sr_demos/demo.py (hunk 7 above)
Comment (georgiablanco) [reply sr_demos/scripts/sr_demos/demo.py]: fixed in https://github.com/shadow-robot/sr_interface/pull/823/commits/4d899740e14c6a37b20146403ea06d3b82a837c8

File: sr_demos/scripts/sr_demos/demo.py (hunk 8 above)
Comment (georgiablanco) [reply sr_demos/scripts/sr_demos/demo.py]: fixed in https://github.com/shadow-robot/sr_interface/pull/823/commits/4d899740e14c6a37b20146403ea06d3b82a837c8

File: sr_demos/scripts/sr_demos/demo.py (hunk 9 above)
Comment (georgiablanco) [reply sr_demos/scripts/sr_demos/demo.py]: fixed in https://github.com/shadow-robot/sr_interface/pull/823/commits/4d899740e14c6a37b20146403ea06d3b82a837c8

File: sr_demos/scripts/sr_demos/demo.py (hunk 10 above)
Comment (georgiablanco) [reply sr_demos/scripts/sr_demos/demo.py]: fixed in https://github.com/shadow-robot/sr_interface/pull/823/commits/4d899740e14c6a37b20146403ea06d3b82a837c8

**Review by bdenoun — APPROVED** [review]

**Review by rnzenha-s — COMMENTED** [review]
Have we tested this code already?

**Comment by georgiablanco** [comment]
> Have we tested this code already?

Yes - as PR template check boxes I have checked both Sim and HW of:
- PST unimanual right
- Biotac bimanual
- Biotact unimanual left
- Simulation no tactiles unimanual left
- Simulation no tactiles unimanual right
- Simulation no tactile bimanual

**Review by rnzenha-s — APPROVED** [review]

**Pull request merged**